File: server/services/workspaceService.js
import { getWorkspaceRecord, upsertWorkspaceRecord } from '../db/client.js';
import { mergeWorkspacePatch, normalizeWorkspace } from '../schemas/workspaceSchema.js';
import { applyEditOperations } from './editValidationService.js';
import { getComponentRegistry } from './modelMetadataService.js';

function nowIso() {
  return new Date().toISOString();
}

function createDefaultWorkspace(projectId) {
  return normalizeWorkspace({
    workspaceId: `workspace_${projectId}`,
    projectId,
    selectedTool: 'mouse',
    rightPanelMode: 'config',
    viewport: {},
    scene: {
      components: [
        {
          id: 'part_main',
          name: 'Main Body',
          visible: true,
          material: { color: '#cccccc', type: 'standard' },
          transform: {
            position: [0, 0, 0],
            rotation: [0, 0, 0],
            scale: [1, 1, 1]
          }
        }
      ]
    },
    lastOperations: [],
    promptHistory: [],
    variantHistory: [],
    history: { past: [], future: [] },
    hasUnsavedOperations: false,
    updatedAt: nowIso()
  });
}

function snapshotWorkspace(workspace) {
  return {
    scene: workspace.scene,
    selectedPartId: workspace.selectedPartId,
    selectedTool: workspace.selectedTool,
    rightPanelMode: workspace.rightPanelMode,
    currentVariantId: workspace.currentVariantId,
    lastOperations: workspace.lastOperations
  };
}

function toRecord(workspace, existingRecord) {
  return {
    id: workspace.workspaceId,
    projectId: workspace.projectId,
    modelId: workspace.modelId,
    currentVariantId: workspace.currentVariantId,
    selectedTool: workspace.selectedTool,
    selectedPartId: workspace.selectedPartId,
    rightPanelMode: workspace.rightPanelMode,
    workspaceStateJson: JSON.stringify(workspace),
    hasUnsavedOperations: workspace.hasUnsavedOperations,
    createdAt: existingRecord?.createdAt ?? workspace.updatedAt,
    updatedAt: workspace.updatedAt
  };
}

async function loadWorkspace(projectId) {
  const existingRecord = await getWorkspaceRecord(projectId);
  if (!existingRecord) {
    return createDefaultWorkspace(projectId);
  }

  try {
    return normalizeWorkspace(JSON.parse(existingRecord.workspaceStateJson));
  } catch {
    return createDefaultWorkspace(projectId);
  }
}

export async function getWorkspace(projectId) {
  if (!projectId) throw new Error('projectId is required.');
  const workspace = await loadWorkspace(projectId);
  const record = await getWorkspaceRecord(projectId);
  if (!record) {
    await upsertWorkspaceRecord(toRecord(workspace));
  }
  return workspace;
}

export async function saveWorkspace(workspace) {
  const normalized = normalizeWorkspace(workspace);
  const existingRecord = await getWorkspaceRecord(normalized.projectId);
  await upsertWorkspaceRecord(toRecord(normalized, existingRecord));
  return normalized;
}

export async function updateWorkspace(projectId, patch) {
  const currentWorkspace = await getWorkspace(projectId);
  const updatedWorkspace = mergeWorkspacePatch(currentWorkspace, patch);
  const withHistory = normalizeWorkspace({
    ...updatedWorkspace,
    history: {
      past: [...currentWorkspace.history.past, snapshotWorkspace(currentWorkspace)].slice(-50),
      future: []
    }
  });
  return saveWorkspace(withHistory);
}

export async function applyWorkspacePrompt(projectId, prompt, operations) {
  const currentWorkspace = await getWorkspace(projectId);
  const componentRegistry = getComponentRegistry(projectId);
  const updatedScene = applyEditOperations(currentWorkspace.scene, operations, {
    components: componentRegistry.components
  });
  const createdAt = nowIso();
  const variantId = `variant_${Date.now()}`;

  return saveWorkspace({
    ...currentWorkspace,
    scene: updatedScene,
    currentVariantId: variantId,
    lastOperations: operations.map((operation) => ({ ...operation, timestamp: createdAt })),
    hasUnsavedOperations: true,
    promptHistory: [
      ...currentWorkspace.promptHistory,
      { id: `prompt_${Date.now()}`, prompt, createdAt }
    ].slice(-50),
    variantHistory: [
      ...currentWorkspace.variantHistory,
      { id: variantId, label: `Variant ${currentWorkspace.variantHistory.length + 1}`, createdAt }
    ].slice(-50),
    history: {
      past: [...currentWorkspace.history.past, snapshotWorkspace(currentWorkspace)].slice(-50),
      future: []
    },
    updatedAt: createdAt
  });
}

export async function undoWorkspace(projectId) {
  const currentWorkspace = await getWorkspace(projectId);
  if (!currentWorkspace.history.past.length) {
    return currentWorkspace;
  }

  const previousSnapshot = currentWorkspace.history.past[currentWorkspace.history.past.length - 1];
  const nextWorkspace = normalizeWorkspace({
    ...currentWorkspace,
    ...previousSnapshot,
    history: {
      past: currentWorkspace.history.past.slice(0, -1),
      future: [...currentWorkspace.history.future, snapshotWorkspace(currentWorkspace)].slice(-50)
    },
    updatedAt: nowIso()
  });
  return saveWorkspace(nextWorkspace);
}

export async function redoWorkspace(projectId) {
  const currentWorkspace = await getWorkspace(projectId);
  if (!currentWorkspace.history.future.length) {
    return currentWorkspace;
  }

  const nextSnapshot = currentWorkspace.history.future[currentWorkspace.history.future.length - 1];
  const nextWorkspace = normalizeWorkspace({
    ...currentWorkspace,
    ...nextSnapshot,
    history: {
      past: [...currentWorkspace.history.past, snapshotWorkspace(currentWorkspace)].slice(-50),
      future: currentWorkspace.history.future.slice(0, -1)
    },
    updatedAt: nowIso()
  });
  return saveWorkspace(nextWorkspace);
}

File: backend/app/services/workspace_service.py
import json
import sqlite3

from app.schemas.workspace import (
    Workspace,
    WorkspacePatch,
    WorkspaceSnapshot,
    create_default_workspace,
    now_iso,
    snapshot_workspace,
)


def save_workspace(db: sqlite3.Connection, workspace: Workspace) -> Workspace:
    existing = db.execute(
        "SELECT created_at FROM workspaces WHERE project_id = ?",
        (workspace.project_id,),
    ).fetchone()
    created_at = existing["created_at"] if existing else workspace.updated_at

    db.execute(
        """
        INSERT INTO workspaces (
          project_id,
          workspace_id,
          model_id,
          current_variant_id,
          selected_tool,
          selected_part_id,
          right_panel_mode,
          workspace_state_json,
          has_unsaved_operations,
          created_at,
          updated_at
        )
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        ON CONFLICT(project_id) DO UPDATE SET
          workspace_id = excluded.workspace_id,
          model_id = excluded.model_id,
          current_variant_id = excluded.current_variant_id,
          selected_tool = excluded.selected_tool,
          selected_part_id = excluded.selected_part_id,
          right_panel_mode = excluded.right_panel_mode,
          workspace_state_json = excluded.workspace_state_json,
          has_unsaved_operations = excluded.has_unsaved_operations,
          updated_at = excluded.updated_at
        """,
        (
            workspace.project_id,
            workspace.workspace_id,
            workspace.model_id,
            workspace.current_variant_id,
            workspace.selected_tool,
            workspace.selected_part_id,
            workspace.right_panel_mode,
            workspace.model_dump_json(by_alias=True),
            int(workspace.has_unsaved_operations),
            created_at,
            workspace.updated_at,
        ),
    )
    db.commit()
    return workspace

def get_workspace(db: sqlite3.Connection, project_id: str) -> Workspace:
    record = db.execute(
        "SELECT workspace_state_json FROM workspaces WHERE project_id = ?",
        (project_id,),
    ).fetchone()
    if not record:
        workspace = create_default_workspace(project_id)
        return save_workspace(db, workspace)

    try:
        return Workspace.model_validate(json.loads(record["workspace_state_json"]))
    except (json.JSONDecodeError, ValueError):
        workspace = create_default_workspace(project_id)
        return save_workspace(db, workspace)


def update_workspace(db: sqlite3.Connection, project_id: str, patch: WorkspacePatch) -> Workspace:
    current = get_workspace(db, project_id)
    patch_data = patch.model_dump(exclude_unset=True)

    viewport_data = current.viewport.model_dump()
    if "viewport" in patch_data and patch_data["viewport"] is not None:
        viewport_patch = patch_data.pop("viewport")
        visible_helpers = viewport_data.get("visible_helpers", {})
        if "visible_helpers" in viewport_patch and viewport_patch["visible_helpers"] is not None:
            visible_helpers = {**visible_helpers, **viewport_patch.pop("visible_helpers")}
        viewport_data = {**viewport_data, **viewport_patch, "visible_helpers": visible_helpers}

    next_data = current.model_dump()
    next_data.update(patch_data)
    next_data["viewport"] = viewport_data
    next_data["project_id"] = current.project_id
    next_data["workspace_id"] = current.workspace_id
    next_data["updated_at"] = now_iso()

    updated = Workspace.model_validate(next_data)
    updated.history.past = [*current.history.past, snapshot_workspace(current)][-50:]
    updated.history.future = []
    return save_workspace(db, updated)


def attach_model_to_workspace(db: sqlite3.Connection, project_id: str, model_id: str) -> Workspace:
    current = get_workspace(db, project_id)
    next_data = current.model_dump()
    next_data["model_id"] = model_id
    next_data["updated_at"] = now_iso()

    updated = Workspace.model_validate(next_data)
    updated.history.past = [*current.history.past, snapshot_workspace(current)][-50:]
    updated.history.future = []
    return save_workspace(db, updated)


def undo_workspace(db: sqlite3.Connection, project_id: str) -> Workspace:
    current = get_workspace(db, project_id)
    if not current.history.past:
        return current

    previous: WorkspaceSnapshot = current.history.past[-1]
    next_data = current.model_dump()
    next_data.update(previous.model_dump())
    next_data["history"] = {
        "past": [item.model_dump() for item in current.history.past[:-1]],
        "future": [
            *[item.model_dump() for item in current.history.future],
            snapshot_workspace(current).model_dump(),
        ][-50:],
    }
    next_data["updated_at"] = now_iso()

    return save_workspace(db, Workspace.model_validate(next_data))


def redo_workspace(db: sqlite3.Connection, project_id: str) -> Workspace:
    current = get_workspace(db, project_id)
    if not current.history.future:
        return current

    next_snapshot: WorkspaceSnapshot = current.history.future[-1]
    next_data = current.model_dump()
    next_data.update(next_snapshot.model_dump())
    next_data["history"] = {
        "past": [
            *[item.model_dump() for item in current.history.past],
            snapshot_workspace(current).model_dump(),
        ][-50:],
        "future": [item.model_dump() for item in current.history.future[:-1]],
    }
    next_data["updated_at"] = now_iso()

    return save_workspace(db, Workspace.model_validate(next_data))
